src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/WhyCommand/WhyCommandArgs.cs

```diff
@@ -33,5 +37,191 @@ internal class WhyCommandArgs
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the list of project paths from the input 'path' argument.
+        /// </summary>
+        /// <returns>List of project paths. Returns null if path was a directory with none or multiple project/solution files.</returns>
+        public IEnumerable<string>? GetListOfProjectPaths()
+        {
+            string fullPath = System.IO.Path.GetFullPath(Path);
+
+            string? projectOrSolutionFile;
+
+            // the path points to a directory
+            if (Directory.Exists(fullPath))
+            {
+                projectOrSolutionFile = GetProjectOrSolutionFileFromDirectory(fullPath);
+
+                if (projectOrSolutionFile == null)
+                {
+                    return null;
+                }
+            }
+            // the path points to a project or solution file
+            else
+            {
+                projectOrSolutionFile = fullPath;
+            }
+
+            return IsSolutionFile(projectOrSolutionFile)
+                        ? MSBuildAPIUtility.GetProjectsFromSolution(projectOrSolutionFile).Where(f => File.Exists(f))
+                        : [projectOrSolutionFile];
+        }
+
+        /// <summary>
+        /// Get the project or solution file from the given directory.
+        /// </summary>
+        /// <returns>A single project or solution file. Returns null if the directory has none or multiple project/solution files.</returns>
+        private string? GetProjectOrSolutionFileFromDirectory(string directory)
+        {
+            var topLevelFiles = Directory.GetFiles(directory, "*.*", SearchOption.TopDirectoryOnly);
+
+            var solutionFiles = topLevelFiles
+                                    .Where(file => IsSolutionFile(file))
+                                    .ToArray();
+            var projectFiles = topLevelFiles
+                                    .Where(file => IsProjectFile(file))
+                                    .ToArray();
+
+            if (solutionFiles.Length + projectFiles.Length > 1)
+            {
+                Logger.LogError(
+                        string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.WhyCommand_Error_MultipleProjectOrSolutionFilesInDirectory,
+                            directory));
+                return null;
+            }
+
+            if (solutionFiles.Length == 1)
+            {
+                return solutionFiles[0];
+            }
+
+            if (projectFiles.Length == 1)
+            {
+                return projectFiles[0];
+            }
+
+            Logger.LogError(
+                string.Format(
+                    CultureInfo.CurrentCulture,
+                    Strings.WhyCommand_Error_NoProjectOrSolutionFilesInDirectory,
+                    directory));
+            return null;
+        }
+
+        private static bool IsSolutionFile(string fileName)
```

Comment:
I'd expect methods like this to already exist for `dotnet list package`, and `dotnet add package`.  Did you search or debug those to find out how they do the same outcome?

While `dotnet nuget why`'s use of MSBuild was sufficiently different to `dotnet list package` to warrant not reusing the same method, this logic of choosing solutions over projects, and finding when there's a single project in a directory vs multiple project files, should be identical across all these commands.

Reply:
I found these implementations in the `dotnet/sdk` repo. `dotnet add package` ([1](https://github.com/dotnet/sdk/blob/main/src/Cli/dotnet/commands/dotnet-add/dotnet-add-package/Program.cs#L29)  -> [2](https://github.com/dotnet/sdk/blob/main/src/Cli/dotnet/MsbuildProject.cs#L78)) only searches for project files in the current directory. `dotnet list package` ([1](https://github.com/dotnet/sdk/blob/main/src/Cli/dotnet/commands/dotnet-list/dotnet-list-package/ListPackageReferencesCommand.cs#L45) -> [2](https://github.com/dotnet/sdk/blob/main/src/Cli/dotnet/commands/dotnet-list/dotnet-list-package/ListPackageReferencesCommand.cs#L68)) searches for a solution file first, and then a project file (same as what we need).

The behavior here matches what those commands do - Search for solution files and project files. Return an error if there are 0 or 2+ project/solution files. Else, return a single project or solution file.

If we want to reuse code rather than writing new code here, I will need to move some more of the parsing for this command to the `dotnet/sdk` repo and then move some of the above methods (for dotnet list package) to more accessible locations within the repo. Would that work better here?

We can also move the code changes in this PR to a more accessible place in NuGet.Client like the [MSBuildAPIUtility](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs) file so that it can be reused in the future.

Reply:
I moved some of these helper methods to XPlatUtility.cs so that they can be reused in the future. I also added unit tests for them.